### PR TITLE
Support equality pushdown on clustering keys for Cassandra

### PR DESCRIPTION
**Pull request by amrutagokhale**
This is still a work in progress. I opened a pull request so as to get some initial feedback from @arhimondr . There are some  `TODO`s in the code that I need to take care of and also test for corner cases. So this isn't a full-fledged implementation, and hence isn't ready for a detailed code review yet.

### CODE REVIEW

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java [hunk 1]
```diff
@@ -196,4 +202,48 @@ public CassandraPartitionResult getPartitions(ConnectorTableHandle tableHandle,
         }
         return Sets.cartesianProduct(partitionColumnValues.build());
     }
+
+    // todo should this be a complete copy of getPartitionKeysSet? for ex., why do you require that the range be a single value"?
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java]: See this: http://www.datastax.com/dev/blog/a-deep-look-to-the-cql-where-clause
The restrictions are different; for partitioning keys, it has to be a single value. For clustering keys, the last one can be a range, but the rest need to be single values.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java [hunk 2]
```diff
@@ -196,4 +202,48 @@ public CassandraPartitionResult getPartitions(ConnectorTableHandle tableHandle,
...
+    private static Set<List<Object>> getClusteringKeysSet(CassandraTable table, TupleDomain<ColumnHandle> tupleDomain)
+    {
+        ImmutableList.Builder<Set<Object>> clusteringColumnValues = ImmutableList.builder();
+        for (CassandraColumnHandle columnHandle : table.getClusteringKeyColumns()) {
+            Domain domain = tupleDomain.getDomains().get().get(columnHandle);
+
+            // if there is no constraint on a clustering key, return an empty set
+            if (domain == null) {
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java]: you don't need a constraint on all of the clustering keys, just a prefix

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java [hunk 3]
```diff
@@ -126,15 +127,20 @@ public CassandraPartitionResult getPartitions(ConnectorTableHandle tableHandle,
...
+        Set<List<Object>> clusteringKeysSet = getClusteringKeysSet(table, tupleDomain);
 
         // empty filter means, all partitions
         if (partitionKeysSet.isEmpty()) {
             return schemaProvider.getAllPartitions(table);
         }
 
         ImmutableList.Builder<ListenableFuture<List<CassandraPartition>>> getPartitionResults = ImmutableList.builder();
+
+        // todo: is the following nested for loop correct?
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java]: As I noted above, I think you should store the clustering key information somewhere other than CassandraPartition; you aren't getting anything out of storing the clustering key info there b/c the underlying Cassandra partitions will be the same. If you're forced to because of how this code is structured, maybe the partitioning key part needs to be refactored...

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java [hunk 4]
```diff
@@ -138,6 +138,7 @@ private static String buildTokenCondition(String tokenExpression, String startTo
         }
         Map<Set<String>, Set<String>> hostsToPartitionKeys = new HashMap<>();
         Map<Set<String>, List<HostAddress>> hostMap = new HashMap<>();
+        Map<Set<String>, String> clusteringPredicateMap = new HashMap<>();
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java]: Can you get around using this? it's hard to follow

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java [hunk 5]
```diff
@@ -88,6 +88,9 @@ public CassandraPartitionResult getPartitions(ConnectorTableHandle tableHandle,
                 List<ColumnHandle> partitionColumns = (List) partitionKeys;
                 remainingTupleDomain = TupleDomain.withColumnDomains(Maps.filterKeys(tupleDomain.getDomains().get(), not(in(partitionColumns))));
                 List<ColumnHandle> clusteringColumns = (List) table.getClusteringKeyColumns();
+                if (!getPartitionKeysSet(table, tupleDomain).isEmpty()) { // Prune clustering keys predicates only if there is a predicate on partition keys too
```
Comment (cawallin) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java]: So you should only add the predicate in the cassandra query if there's a predicate on partition keys too

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CachingCassandraSchemaProvider.java [hunk 6]
```diff
@@ -287,7 +288,7 @@ private CassandraTable loadTable(final SchemaTableName tableName)
     {
         return retry()
                 .stopOnIllegalExceptions()
-                .run("getPartitions", () -> session.getPartitions(key.getTable(), key.getFilterPrefix()));
+                .run("getPartitions", () -> session.getPartitions(key.getTable(), key.getFilterPrefix(), key.getClusteringKeysFilterPrefix()));
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CachingCassandraSchemaProvider.java]: +1

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 7]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
     {
         CassandraTableHandle handle = checkType(table, CassandraTableHandle.class, "table");
         CassandraPartitionResult result = partitionManager.getPartitions(handle, constraint.getSummary());
-        ConnectorTableLayout layout = getTableLayout(session, new CassandraTableLayoutHandle(handle, result.getPartitions()));
+        ConnectorTableLayout layout = getTableLayout(session, new CassandraTableLayoutHandle(handle, result.getPartitions(), buildClusteringPredicate(schemaProvider.getTable(handle), constraint)));
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: I think you don't need to build clustering key predicate when it is unpartitioned full-scan.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 8]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
         return ImmutableList.of(new ConnectorTableLayoutResult(layout, result.getUnenforcedConstraint()));
     }
 
+    private static Optional<String> buildClusteringPredicate(CassandraTable cassandraTable, Constraint<ColumnHandle> constraint)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: Pass `constraint.getSummary()` here instead of `Constraint<ColumnHandle>`

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTable.java [hunk 9]
```diff
@@ -49,6 +51,31 @@ public CassandraTableHandle getTableHandle()
...
     }
 
+    public List<CassandraColumnHandle> getClusteringKeyColumns()
+    {
+        return columns.stream()
+                .filter(CassandraColumnHandle::isClusteringKey)
+                .collect(toList());
+    }
+
+    public boolean isLastClusteringKeyColumn(CassandraColumnHandle columnHandle)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTable.java]: Do we really need this method? Can't we assume that clustering columns returned by `getClusteringKeyColumns` is in correct order? And check it in the loop that it is last element?

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 10]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+
+    private static Set<List<Object>> getClusteringKeysSet(CassandraTable table, TupleDomain<ColumnHandle> tupleDomain)
+    {
+        boolean predicateMissing = false;
+        ImmutableList.Builder<Set<Object>> clusteringColumnValues = ImmutableList.builder();
+        for (CassandraColumnHandle columnHandle : table.getClusteringKeyColumns()) {
+            Domain domain = tupleDomain.getDomains().get().get(columnHandle);
+
+            if (domain == null) {
+                predicateMissing = true; // acknowledge that predicate on a clustering key is absent
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: i think you can just brake the loop here. No need to go forward.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 11]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+                continue;
+            }
+            else if (domain != null && predicateMissing) { // clustering keys predicate is not a prefix
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: This won't be needed then

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 12]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+                return ImmutableSet.of();
+            }
+
+            // todo does cassandra allow null clustering keys?
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: As soon as i know - it doesn't. Remove the comment.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 13]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+            if (domain.isNullAllowed()) {
+                return ImmutableSet.of();
+            }
+
+            Set<Object> values = domain.getValues().getValuesProcessor().transform(
+                    ranges -> {
+                        ImmutableSet.Builder<Object> columnValues = ImmutableSet.builder();
+                        for (Range range : ranges.getOrderedRanges()) {
+                            // Prior to Cassandra 2.2, only the last clustering column can have a range of values
+                            if (!range.isSingleValue() && !table.isLastClusteringKeyColumn(columnHandle)) {
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]:
```
if(range.isSingleValue()){
Object value = range.getSingleValue();
CassandraType valueType = columnHandle.getCassandraType();  columnValues.add(valueType.validateClusteringKey(value));
} else if (isLastClusteringKeyColumn){
// try to handle whatever
}
```

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 14]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+                                return ImmutableSet.of();
+                            }
+                            else if (!range.isSingleValue() && table.isLastClusteringKeyColumn(columnHandle)) {
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: Remove this `isLastClusteringKeyColumn` method. Use just `for(;;)` loop. And determine whether it is last clustering key by the index.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 15]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+
+                stringBuilder.append(CassandraCqlUtils.validColumnName(clusteringKeyColumns.get(i).getName()));
+                stringBuilder.append(" = ");
+                stringBuilder.append(CassandraCqlUtils.cqlValue(toCQLCompatibleString(clusteringKeys.get(i)), clusteringKeyColumns.get(i).getCassandraType()));
+            }
+        }
+        return Optional.of(stringBuilder.toString());
+    }
+
+    private static Set<List<Object>> getClusteringKeysSet(CassandraTable table, TupleDomain<ColumnHandle> tupleDomain)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: Return `List<List<Object>>` from here

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 16]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+                    discreteValues -> {
+                        if (discreteValues.isWhiteList()) {
+                            return ImmutableSet.copyOf(discreteValues.getValues());
+                        }
+                        return ImmutableSet.of();
+                    },
+                    allOrNone -> ImmutableSet.of());
+            clusteringColumnValues.add(values);
+        }
+        return Sets.cartesianProduct(clusteringColumnValues.build());
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: static import

Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: ImmutableList.copyOf(cartesianProduct(clusteringColumnValues.build()))

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 17]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
-        ConnectorTableLayout layout = getTableLayout(session, new CassandraTableLayoutHandle(handle, result.getPartitions()));
+        ConnectorTableLayout layout = getTableLayout(session, new CassandraTableLayoutHandle(handle, result.getPartitions(), buildClusteringPredicate(schemaProvider.getTable(handle), constraint)));
         return ImmutableList.of(new ConnectorTableLayoutResult(layout, result.getUnenforcedConstraint()));
     }
 
+    private static Optional<String> buildClusteringPredicate(CassandraTable cassandraTable, Constraint<ColumnHandle> constraint)
+    {
+        List<CassandraColumnHandle> clusteringKeyColumns = cassandraTable.getClusteringKeyColumns();
+        Set<List<Object>> clusteringKeysSet = getClusteringKeysSet(cassandraTable, constraint.getSummary());
+        if (clusteringKeysSet == null || clusteringKeysSet.isEmpty()) {
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: `getClusteringKeysSet`  doesn't return `null`. Just check whether it is not empty.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java [hunk 18]
```diff
@@ -198,10 +204,88 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
...
+            return Optional.empty();
+        }
+
+        // todo Is it correct the way predicates are being appended to each other?
+        StringBuilder stringBuilder = new StringBuilder();
+        for (int j = 0; j < clusteringKeysSet.size(); j++) {
+            List<Object> clusteringKeys = clusteringKeysSet.iterator().next();
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: It will always return the very first element. I think that is not what you actually want to do. Once you change `getClusteringKeysSet` to return `List` - you can just do `clusteringKeysSet.get(j)`

Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: You should concat cartesian product tuples with `OR`. Check if it is supported by Cassandra.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java [hunk 19]
```diff
@@ -87,6 +88,17 @@ public CassandraPartitionResult getPartitions(ConnectorTableHandle tableHandle,
                 @SuppressWarnings({"rawtypes", "unchecked"})
                 List<ColumnHandle> partitionColumns = (List) partitionKeys;
                 remainingTupleDomain = TupleDomain.withColumnDomains(Maps.filterKeys(tupleDomain.getDomains().get(), not(in(partitionColumns))));
+                List<CassandraColumnHandle> clusteringColumns = table.getClusteringKeyColumns();
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraPartitionManager.java]: I don't think you need to prune clustering predicates at this step. It must be prunned in a place where you actually build the clustering predicate. And it must be prunned from the `unenforced predicates` returned from the partition manager.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java [hunk 20]
```diff
@@ -176,7 +180,7 @@ private static String buildTokenCondition(String tokenExpression, String startTo
                     size++;
                     if (size > partitionSizeForBatchSelect) {
                         String partitionId = String.format("%s in (%s)", partitionKeyColumnName, sb.toString());
-                        CassandraSplit split = new CassandraSplit(connectorId, schema, table, partitionId, null, hostMap.get(entry.getKey()));
+                        CassandraSplit split = new CassandraSplit(connectorId, schema, table, partitionId, clusteringKeysPredicate, hostMap.get(entry.getKey()));
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java]: Make sure that cassandra supports queries like `partition_key IN <value> AND clustering_key = <value>`

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTable.java [hunk 21]
```diff
@@ -49,6 +51,31 @@ public CassandraTableHandle getTableHandle()
...
+    {
+        List<CassandraColumnHandle> clusteringKeyColumns = getClusteringKeyColumns();
+        return (clusteringKeyColumns.get(clusteringKeyColumns.size() - 1).equals(columnHandle));
+    }
+
+    public boolean isAPrefix(Set<CassandraColumnHandle> columnHandles)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTable.java]: I don't think you would need this method once you refactor `unenforcedPredicates`

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTableLayoutHandle.java [hunk 22]
```diff
@@ -28,22 +28,24 @@
 {
     private final CassandraTableHandle table;
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTableLayoutHandle.java]: This class is really error pron. The serialization roundtrip must yield the same object. It is not true for this class. Please refactor it. It shouldn't be hard. And add json serialization round trip unit test.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTableLayoutHandle.java [hunk 23]
```diff
@@ -28,22 +28,24 @@
...
     private final Optional<List<CassandraPartition>> partitions;
+    private final Optional<String> clusteringKeysPredicate;
 
     @JsonCreator
     public CassandraTableLayoutHandle(@JsonProperty("table") CassandraTableHandle table)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTableLayoutHandle.java]: Leave only single constructor

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTableLayoutHandle.java [hunk 24]
```diff
@@ -58,6 +60,12 @@ public CassandraTableHandle getTable()
         return partitions;
     }
 
+    @JsonIgnore
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraTableLayoutHandle.java]: Ignore nothing

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java (hunk 8 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: It would be nice to add unit tests for this method. Maybe you can extract this algorithm to a separate class, and then unit-test it.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java (hunk 20 above)
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java]: Try to cover all the corner cases with a product/integration test.

like

- `partition_key IN <value> AND clustering_key = <value>`
- `partition_key IN <value> AND (clustering_key = <value> OR clustering_key = <value> )`
-  `partition_key1 = <value> AND partition_key2 = <value> AND ((clustering_key1 = <value> AND clustering_key2 = <value> OR (clustering_key1 = <value> AND clustering_key2 = <value>) OR (clustering_key1 = <value>))`
- etc.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java (hunk 18 above)
Comment (amrutagokhale) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraMetadata.java]: Looks like Cassandra doesn't support `OR`. So for now, we may have to forgo pushing down such predicates.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/ClusteringKeysUtils.java [hunk 25]
```diff
@@ -0,0 +1,131 @@
...
+                stringBuilder.append(" = ");
+                stringBuilder.append(CassandraCqlUtils.cqlValue(toCQLCompatibleString(columnValues.get(i)), clusteringKeyColumns.get(i).getCassandraType()));
+            }
+            clusteringPredicates.add(stringBuilder.toString());
+        }
+        return Optional.of(clusteringPredicates.build());
+    }
+
+    // return clustering key columns whose predicates are to be pushed down
+    public static TupleDomain<ColumnHandle> pushedDownClusteringColumns(CassandraTable table, TupleDomain<ColumnHandle> tupleDomain)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/ClusteringKeysUtils.java]: pushedDownClusteringPredicates

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/util/TestClusteringKeys.java [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static org.testng.Assert.assertEquals;
+
+public class TestClusteringKeys
+{
+    @Test
+    public void testBuildClusteringPredicate()
```
Comment (arhimondr) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/util/TestClusteringKeys.java]: Add test for the other method from the utility class

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/util/TestClusteringKeys.java [hunk 27]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static org.testng.Assert.assertEquals;
+
+public class TestClusteringKeys
```
Comment (arhimondr) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/util/TestClusteringKeys.java]: Keep the test name consistent. It should follow the `Test<ClassName>` pattern.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/ClusteringKeysUtils.java [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.facebook.presto.cassandra.util.CassandraCqlUtils.toCQLCompatibleString;
+import static com.google.common.collect.Sets.cartesianProduct;
+
+public final class ClusteringKeysUtils
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/ClusteringKeysUtils.java]: Make this class non static. Call it `ClusteringKeyPredicatesExtractor`

Pass `CassandraTable` and `TupleDomain<ColumnHandle> unenforcedPredicates` to the constructor. 

`unenforcedPredicates` - are the predicated returned from the partition manager.

This class should have 2 methods:

- `List<String> getClusteringKeyPredicates()`
- `TupleDomain<ColumnHandle> getUnenforcedPredicates()`

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java [hunk 29]
```diff
@@ -176,17 +186,21 @@ private static String buildTokenCondition(String tokenExpression, String startTo
                     size++;
                     if (size > partitionSizeForBatchSelect) {
                         String partitionId = String.format("%s in (%s)", partitionKeyColumnName, sb.toString());
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraSplitManager.java]: Make sure this branch works.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+
+import static com.facebook.presto.cassandra.util.CassandraCqlUtils.toCQLCompatibleString;
+import static com.google.common.collect.Sets.cartesianProduct;
+import static java.util.Objects.requireNonNull;
+
+public class CassandraClusteringPredicatesExtractor
+{
+    private final List<CassandraColumnHandle> clusteringColumns;
+    private final TupleDomain<ColumnHandle> predicates;
+    private ClusteringPushDownResult clusteringPushDownResult;
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: final

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 31]
```diff
@@ -0,0 +1,158 @@
...
+        Map<ColumnHandle, Domain> notPushedDown = new HashMap<>(predicates.getDomains().get());
+
+        if (!notPushedDown.isEmpty() && !pushedDown.isEmpty()) {
+            notPushedDown.entrySet().removeAll(pushedDown.entrySet());
+        }
+
+        return TupleDomain.withColumnDomains(notPushedDown);
+    }
+
+    private static class ClusteringPushDownResult
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: Make this class to be pojo. Extrac logic to the static method.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 32]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        private Map<ColumnHandle, Domain> domains;
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: final

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 33]
```diff
@@ -0,0 +1,158 @@
...
+        private Set<List<Object>> domainValues;
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: final

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 34]
```diff
@@ -0,0 +1,158 @@
...
+        {
+            return domains;
+        }
+
+        public ClusteringPushDownResult(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates)
+        {
+            domainValues = getClusteringKeysSet(clusteringColumns, predicates);
+        }
+
+        private Set<List<Object>> getClusteringKeysSet(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates)
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: extract this method. Make it return `ClusteringPushDownResult`

File: presto-cassandra/src/test/java/com/facebook/presto/cassandra/TestCassandraIntegrationSmokeTest.java [hunk 35]
```diff
@@ -121,6 +122,33 @@ public void testCreateTableAs()
         execute("DROP TABLE table_all_types_copy");
     }
 
+    @Test
```
Comment (arhimondr) [inline presto-cassandra/src/test/java/com/facebook/presto/cassandra/TestCassandraIntegrationSmokeTest.java]: Add test cases for the table with multiple partitions key and multiple clustering keys

File: presto-main/etc/config.properties [hunk 36]
```diff
@@ -36,6 +36,7 @@ plugin.bundles=\
   ../presto-tpch/pom.xml, \
   ../presto-local-file/pom.xml, \
   ../presto-mysql/pom.xml,\
+  ../presto-cassandra/pom.xml,\
```
Comment (arhimondr) [inline presto-main/etc/config.properties]: Rebase onto actual master

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 37]
```diff
@@ -0,0 +1,159 @@
...
+    public CassandraClusteringPredicatesExtractor(List<CassandraColumnHandle> clusteringColumns, TupleDomain<ColumnHandle> predicates)
+    {
+        this.clusteringColumns = ImmutableList.copyOf(requireNonNull(clusteringColumns, "clusteringColumns is null"));
+        this.predicates = requireNonNull(predicates, "predicates is null");
+        this.clusteringPushDownResult = getClusteringKeysSet(clusteringColumns, predicates);
+    }
+
+    public List<String> getClusteringKeyPredicates()
+    {
+        final Set<List<Object>> pushedDownDomainValues = clusteringPushDownResult.getDomainValues();
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: remove final

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 38]
```diff
@@ -0,0 +1,159 @@
...
+            for (int i = 0; i < clusteringKeys.size(); i++) {
+                if (i > 0) {
+                    stringBuilder.append(" AND ");
+                }
+
+                stringBuilder.append(CassandraCqlUtils.validColumnName(clusteringColumns.get(i).getName()));
+                stringBuilder.append(" = ");
+                stringBuilder.append(CassandraCqlUtils.cqlValue(toCQLCompatibleString(clusteringKeys.get(i)), clusteringColumns.get(i).getCassandraType()));
+            }
+            if (!stringBuilder.toString().isEmpty() && !stringBuilder.toString().equals("")) {
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: suggestion: instead of doing this ugly check do `continue` at the beginning of the loop if the clustering keys are empty.

File: presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java [hunk 39]
```diff
@@ -0,0 +1,159 @@
...
+                    },
+                    discreteValues -> {
+                        if (discreteValues.isWhiteList()) {
+                            return ImmutableSet.copyOf(discreteValues.getValues());
+                        }
+                        return ImmutableSet.of();
+                    },
+                    allOrNone -> ImmutableSet.of());
+
+            clusteringColumnValues.add(values);
```
Comment (arhimondr) [inline presto-cassandra/src/main/java/com/facebook/presto/cassandra/CassandraClusteringPredicatesExtractor.java]: if `values` is empty, does it make sense to add it to the `clusteringColumnValues`? shouldn't it be within the `if` clause?

**Comment by sumitkgec** [comment]
This is a really a big milestone for Cassandra presto connector. Do you guys have any timeline when it will be merged and released?

I can see predict push down for clustering key is missing. Which really impacting the performance for Presto. It taking entire partition data and doing the calculation, this can be faster if Presto will pass the partition and primary key to Cassandra and do the calculation on the small portion of data.

This connector can be just like Datastax Spark-Cassandra connector. Cassandra with Presto will be a very good solution.

**Comment by amrutagokhale** [comment]
@sumitkgec We will try to get this merged some time soon!

**Comment by cawallin** [comment]
Merged, thanks!

**Comment by sumitkgec** [comment]
Thanks a lot!! 

I have one question - Wheather predict Push down on clustering keys to Cassandra from Presto will work for timestamp date type(between or > or >= or < or <= syntax)?

**Comment by amrutagokhale** [comment]
The merged patch supports pushdown only for equality predicates (= or IN clause) but doesn't support pushdown for inequality predicates (> or >= or < or <= or !=). If you would like to open a PR though, we would be happy to review it!

**Comment by sumitkgec** [comment]
@amrutagokhale Thanks a lot. Cassandra allows inequality restrictions(>, >=, <=, <) only on the last clustering column being restricted. I will try to open the PR changes for this sometime soon.

Just for reference - Cassandra allow below restriction in WHERE clause for SELECT statements(except Secondary Index)
**Partition keys restrictions**
The partition key columns support only two operators(except token function): = and IN.
**Clustering column restrictions**
Clustering columns support the =, IN, >, >=, <=, < restrictions. 

Presto already supports partition key = and IN operators push down to Cassandra and from next release, it will support clustering keys = and IN operators push down. Let me know if I have missed anything here so that I will try to add this in PR changes.

**Comment by amrutagokhale** [comment]
That sounds right, @sumitkgec.

**Comment by sumitkgec** [comment]
@amrutagokhale : I have opened the PR to support the complete predicate pushdown for clustering keys.

[https://github.com/prestodb/presto/pull/7833](url) 

Could you please review the code and provide your initial feedback.
